### PR TITLE
[programmable transactions] Publish produces UpgradeCap as result

### PR DESCRIPTION
**Pull request by amnn**
## Description 

When a package is successfully published in a programmable transaction, produce an `UpgradeCap` which gives its owner the ability to upgrade the just-published package.

## Test Plan 

TBD, waiting on tests to migrate to using programmable transactions, at which point, we can test that publish produces this new object.

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [x] user-visible impact
- [ ] breaking change for a client SDKs
- [x] breaking change for FNs (FN binary must upgrade)
- [x] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes

Publish events now return an `UpgradeCap` as their command result, on success, which gives its owner the ability to upgrade the published package.  Note that this value cannot be dropped, so it must either be transfered, or explicitly burned using `public entry fun sui::packge::make_immutable(cap: UpgradeCap)`.

### CODE REVIEW

File: crates/sui-framework/sources/package.move
Comment (amnn) [inline crates/sui-framework/sources/package.move:31]: I also "spaced out" the constants for upgrade policies, in case we need to add more after the fact, but without losing the nice property that higher numbers are more restrictive.

File: crates/sui-adapter/src/programmable_transactions/execution.rs
Comment (tnowacki) [inline crates/sui-adapter/src/programmable_transactions/execution.rs:404]: I am very glad to see this was easy to do! (I hope! lol)

File: crates/sui-types/src/programmable_transaction_builder.rs
Comment (tnowacki) [inline crates/sui-types/src/programmable_transaction_builder.rs:172]: publish + make immutable feels like the most logical choice for easy migration. 

But thoughts on something like `publish_and_take_cap`, which transfers it to the sender. But that got me thinking... maybe we need something like `Argument::Sender` and maybe `Argument::Sponsor`? Since it's actually quite difficult to add something that transfers the publish cap to the sender, without hard coding the sender address as an argument... which would also needlessly increase the txn size. 

Thoughts?

**Review by tnowacki — COMMENTED** [review]
Overall looks great!

**Review by rvantonder — APPROVED** [review]
walked through this live with @amnn, looks good to me!

File: crates/sui-types/src/programmable_transaction_builder.rs
Comment (amnn) [reply crates/sui-types/src/programmable_transaction_builder.rs:172]: +1 `Argument::sender` at least -- that would help a lot. cc @tzakian as this will be relevant to you creating a test that sends the cap back to sender!

Comment (tzakian) [reply crates/sui-types/src/programmable_transaction_builder.rs:172]: Bit of a drive-by comment, but I was just looking at this exact piece of code and thinking the exact same thing you are with having a `publish_and_take_cap`. (for context: I'm working on enabling grabbing the UpgradeCap in transactional tests ATM). I was running into the exact issue you're mentioning of this would then lead to hard-coding the sender address in the `Command` which isn't all that great..

Having `Argument::Sender` (at least) would make sense to me, and `Sponsor` to some extent (although not as much as `Sender` IMO).

**Review by tnowacki — COMMENTED** [review]
You will probably wan't to wait until #8881 and #8993 both land before landing this, to get a good bit of coverage from tests.

File: crates/sui-types/src/programmable_transaction_builder.rs
Comment (amnn) [reply crates/sui-types/src/programmable_transaction_builder.rs:172]: Great minds!

File: crates/sui-types/src/move_package.rs
Comment (ronny-mysten) [inline crates/sui-types/src/move_package.rs:206]:
```suggestion
    /// the fresh `uid`.
```

File: crates/sui-adapter/src/programmable_transactions/execution.rs
Comment (amnn) [reply crates/sui-adapter/src/programmable_transactions/execution.rs:404]: It was!

File: sdk/typescript/src/builder/legacy.ts
Comment (amnn) [inline sdk/typescript/src/builder/legacy.ts:135]: Heads up @666lcz, I ported the change I made to the programmable transaction builder in Rust to the migration logic from legacy transactions -- a legacy transaction now drops the `UpgradeCap` that now gets produced.

**Pull request merged**
